### PR TITLE
 [pulsar admin] getPartitionedTopicMetada method support setting auto create topic

### PR DESCRIPTION
**Pull request by shibd**
#11923 

### Motivation

https://github.com/streamnative/pulsar-flink/issues/398

### Modifications

1.  Add validateTopicPolicyOperation to PersistentTopic#getPartitionedMetadata
2.  `Topics#getPartitionedTopicMetadata` add parameter `checkAllowAutoCreation`

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
- `PersistentTopicsTest#testGetPartitionedMetadata`

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: **yes**
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: **yes**
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

Need to update docs? 

- [x] doc-required 
- [ ] no-need-doc 
- [ ] doc

### CODE REVIEW

**Comment by shibd** [comment]
@codelipenghui Can you help me review it?

**Comment by Anonymitaet** [comment]
@shibd Thanks for your contribution. For this PR, do we need to update docs?

(The [PR template contains info about doc](https://github.com/apache/pulsar/blob/master/.github/PULL_REQUEST_TEMPLATE.md#documentation), which helps others know more about the changes. Can you provide doc-related info in this and future PR descriptions? Thanks)

**Comment by shibd** [comment]
@Anonymitaet  Thanks,I readjusted the PR description.

**Comment by Anonymitaet** [comment]
@shibd Thanks for your contribution. Please do not forget to add docs accordingly to allow users to know your great code changes. And you can ping me to review the docs, thanks.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 1]
```diff
@@ -786,6 +786,9 @@ public PartitionedTopicMetadata getPartitionedMetadata(
             @ApiParam(value = "Is check configuration required to automatically create topic")
             @QueryParam("checkAllowAutoCreation") @DefaultValue("false") boolean checkAllowAutoCreation) {
         validateTopicName(tenant, namespace, encodedTopic);
+        if (checkAllowAutoCreation) {
```
[the remark this answers is not in the document]
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: This method can auto create topic, so need  validate `WRITE` permission.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 2]
```diff
@@ -389,6 +392,27 @@ public void testNonPartitionedTopics() {
...
+        // 2. Add validate policy assert.
+        doThrow(new RestException(Response.Status.FORBIDDEN, "mock message"))
+                .when(persistentTopics).validateTopicPolicyOperation(persistentTopics.topicName, PolicyName.PARTITION, PolicyOperation.WRITE);
+        try {
+            persistentTopics.getPartitionedMetadata(testTenant, testNamespace, testLocalTopicName, true, true);
+            Assert.fail("Should fail validation on policy exception");
+        } catch (RestException e) {
+            Assert.assertEquals(Response.Status.FORBIDDEN.getStatusCode(), e.getResponse().getStatus());
+        }
+
```
Comment (gaoran10) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:413]: Could we add tests to when the param `checkAllowAutoCreation` is `false`?

Comment (shibd) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java]: Good idea, I added.

**Comment by shibd** [comment]
@Anonymitaet  I added docs, Can you help me review it?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarvinCai) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: then what if user with only read permission just want to get meta data? they should be able to fetch the partition meta right.

[the remark this answers is not in the document]
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: Yes, Validate `WRITE` permission only when `checkAllowAutoCreation  = true`.  I added this unit test case.

[the remark this answers is not in the document]
Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: @MarvinCai Hi, Do you have any more suggestions?

**Comment by MarvinCai** [comment]
cpp test is broke with this patch, can you take a look:
```
2021-09-15 03:10:19.336 INFO  [140681233118976] HTTPLookupService:238 | Curl Lookup Request sent for http://localhost:8080/admin/v2/persistent/private/auth/test-token-http/partitions?checkAllowAutoCreation=true
2021-09-15 03:10:19.344 ERROR [140681233118976] HTTPLookupService:263 | Response failed for url http://localhost:8080/admin/v2/persistent/private/auth/test-token-http/partitions?checkAllowAutoCreation=true. Error Code 22
2021-09-15 03:10:19.344 ERROR [140681233118976] ClientImpl:188 | Error Checking/Getting Partition Metadata while creating producer on persistent://private/auth/test-token-http -- ConnectError
/pulsar/pulsar-client-cpp/tests/AuthTokenTest.cc:128: Failure
```

**Comment by shibd** [comment]
> cpp test is broke with this patch, can you take a look:

Thank reply, I fix it. Can you take look?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java [hunk 3]
```diff
@@ -786,6 +786,9 @@ public PartitionedTopicMetadata getPartitionedMetadata(
             @ApiParam(value = "Is check configuration required to automatically create topic")
             @QueryParam("checkAllowAutoCreation") @DefaultValue("false") boolean checkAllowAutoCreation) {
         validateTopicName(tenant, namespace, encodedTopic);
+        if (authoritative && checkAllowAutoCreation) {
```
Comment (MarvinCai) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: I don't think we should include `authoritive` here? https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/LookupOptions.java#L28-L31
https://github.com/apache/pulsar/blame/master/site2/website/versioned_docs/version-2.8.1/developing-binary-protocol.md#L478-L480

Comment (shibd) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/PersistentTopics.java:789]: sorry, I got it wrong. I changed it back.

**Comment by MarvinCai** [comment]
@shibd the cpp test should be able to fixed by adding `-r "token-principal"` to line  110 of `pulsar-test-service-start.sh`

**Comment by MarvinCai** [comment]
but I think this pr is introducing a breaking change that now producer will need admin permission to be able to produce if using HttpLookupService, since when creating producer it'll get topic metadata first, which now validate `WRITE` permission on `TOPIC`, while current authz check is coarse-grained so it just check if role has admin permission
that's why the cpp test is failing
@codelipenghui @hangc0276 @eolivelli any thought on this?

**Comment by eolivelli** [comment]
> but I think this pr is introducing a breaking change that now producer will need admin permission to be able to produce if using HttpLookupService, since when creating producer it'll get topic metadata first, which now validate `WRITE` permission on `TOPIC`, while current authz check is coarse-grained so it just check if role has admin permission
> that's why the cpp test is failing
> @codelipenghui @hangc0276 @eolivelli any thought on this?

I recently filed this issue for a similar problem
https://github.com/apache/pulsar/issues/11945

btw you are right, this is a breaking change.

I believe that we should relax the authz checks if you only want to get the list of partitions, you do not need "admin" access

**Comment by eolivelli** [comment]
> @shibd the cpp test should be able to fixed by adding `-r "token-principal"` to line 110 of `pulsar-test-service-start.sh`

@MarvinCai can you please send a patch ?

**Comment by MarvinCai** [comment]
I think it's safer to just paste the patch as it's just single line change:
```
diff --git a/pulsar-client-cpp/pulsar-test-service-start.sh b/pulsar-client-cpp/pulsar-test-service-start.sh
index 248d628b9c2..2b656ddf6dd 100755
--- a/pulsar-client-cpp/pulsar-test-service-start.sh
+++ b/pulsar-client-cpp/pulsar-test-service-start.sh
@@ -107,7 +107,7 @@ $PULSAR_DIR/bin/pulsar-admin namespaces grant-permission public/default-4 \
 $PULSAR_DIR/bin/pulsar-admin namespaces set-encryption-required public/default-4 -e
 
 # Create "private" tenant
-$PULSAR_DIR/bin/pulsar-admin tenants create private -r "" -c "standalone"
+$PULSAR_DIR/bin/pulsar-admin tenants create private -r "token-principal" -c "standalone"
 
 # Create "private/auth" with required authentication
 $PULSAR_DIR/bin/pulsar-admin namespaces create private/auth --clusters standalone

```

**Comment by eolivelli** [comment]
@MarvinCai 
in order to apply the patch we need a PR, this is our merge workflow.
if I create the PR I will be listed as author, but you are the actual author of the patch, so it would be very better that you post the PR

**Comment by shibd** [comment]
@eolivelli

**Comment by shibd** [comment]
@MarvinCai @eolivelli  hi, I fixed cpp unit test, can you help me review it?

**Comment by syhily** [comment]
@BewareMyPower This issue block the pulsar-flink-connector. Can you help me review this MR?

**Review by BewareMyPower — APPROVED** [review]

**Comment by BewareMyPower** [comment]
This PR is too late to review. I'm afraid it could not be included in Pulsar 2.9.0 now.

This PR adds a new overridden of `getPartitionedTopicMetadataAsync` to `Topics`, but it doesn't affect the existing APIs. Is it suitable to add this PR to 2.8.2? @315157973

**Comment by BewareMyPower** [comment]
@MarvinCai @gaoran10 PTAL again. I think the comments are all addressed now.

File: site2/docs/admin-api-topics.md [hunk 4]
```diff
@@ -1138,12 +1138,18 @@ $ pulsar-admin topics get-partitioned-topic-metadata \
 ```
 
 <!--REST API-->
```
Comment (Anonymitaet) [inline site2/docs/admin-api-topics.md:1140]: Here we have `<!--REST API-->`  and `<!--Java-->` methods, do we have `<!--pulsar-admin-->` method?

File: site2/docs/admin-api-topics.md [hunk 5]
```diff
@@ -1138,12 +1138,18 @@ $ pulsar-admin topics get-partitioned-topic-metadata \
...
+When topic auto-creation is enabled by broker, you can set param `checkAllowAutoCreation = true` to auto create topic.
```
Comment (Anonymitaet) [inline site2/docs/admin-api-topics.md:1141]:
```suggestion
When enabling topic auto-creation, you can set the parameter `checkAllowAutoCreation = true` to automatically create a topic.
```

do you mean this? if so, please help update all occurrences.

File: site2/docs/admin-api-topics.md (hunk 4 above)
Comment (shibd) [reply site2/docs/admin-api-topics.md:1140]: No,  pulsar-admin does not have this method

**Pull request closed without merging**